tests/units/units.cpp

```diff
@@ -2799,3 +2831,593 @@ TEST(Units, equivalentUnitsMatchingBuiltinUnitsFromVariable)
+    ic->setImportReference("my_component");
+
+    model->addComponent(ic);
+
+    auto flatModel = importer->flattenModel(model);
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+}
+
+TEST(Units, flattenImportedUnitsWithNonStandardReferenceDirectlyWithExistingEquivalentUnits)
+{
+    libcellml::ModelPtr importModel;
+    libcellml::ImporterPtr importer = libcellml::Importer::create();
+    libcellml::ImportSourcePtr import = createBaseModelImport(importModel, "importer/units/base_model.cellml");
+
+    libcellml::ModelPtr model = libcellml::Model::create("main_model");
+
+    libcellml::UnitsPtr iu = libcellml::Units::create("per_fmol");
+    iu->setImportSource(import);
+    iu->setImportReference("per_fmol");
+
+    libcellml::UnitsPtr u = libcellml::Units::create("fmol");
+    u->addUnit("mole", "femto");
+    model->addUnits(iu);
+    model->addUnits(u);
+
+    auto flatModel = importer->flattenModel(model);
+
+    EXPECT_EQ(size_t(2), flatModel->unitsCount());
+    EXPECT_EQ("per_fmol", flatModel->units(0)->name());
+    EXPECT_EQ("fmol", flatModel->units(1)->name());
+}
+
+TEST(Units, flattenImportedUnitsWithNonStandardReferenceDirectlyWithExistingEquivalentUnitsWithDifferentName)
+{
+    libcellml::ModelPtr importModel;
+    libcellml::ImporterPtr importer = libcellml::Importer::create();
+    libcellml::ImportSourcePtr import = createBaseModelImport(importModel, "importer/units/base_model.cellml");
+
+    libcellml::ModelPtr model = libcellml::Model::create("main_model");
+
+    libcellml::UnitsPtr iu = libcellml::Units::create("per_fmol");
+    iu->setImportSource(import);
+    iu->setImportReference("per_fmol");
+
+    libcellml::UnitsPtr u = libcellml::Units::create("femto_mole");
+    u->addUnit("mole", "femto");
+    model->addUnits(iu);
+    model->addUnits(u);
+
+    auto flatModel = importer->flattenModel(model);
+
+    EXPECT_EQ(size_t(2), flatModel->unitsCount());
+    EXPECT_EQ("per_fmol", flatModel->units(0)->name());
+    EXPECT_EQ("femto_mole", flatModel->units(1)->name());
+}
+
+TEST(Units, flattenImportedUnitsWithNonStandardReferenceDirectlyWithExistingEquivalentUnitsThroughComponent)
+{
+    libcellml::ModelPtr importModel;
+    libcellml::ImporterPtr importer = libcellml::Importer::create();
+    libcellml::ImportSourcePtr import = createBaseModelImport(importModel, "importer/units/base_model.cellml");
+
+    libcellml::ModelPtr model = libcellml::Model::create("main_model");
+
+    libcellml::ComponentPtr ic = libcellml::Component::create("my_component");
+    ic->setImportSource(import);
+    ic->setImportReference("my_component");
+
+    libcellml::UnitsPtr u = libcellml::Units::create("fmol");
+    u->addUnit("mole", "femto");
+
+    model->addComponent(ic);
+    model->addUnits(u);
+
+    auto flatModel = importer->flattenModel(model);
+
+    EXPECT_EQ(size_t(3), flatModel->unitsCount());
+    EXPECT_EQ("fmol", flatModel->units(0)->name());
+    EXPECT_EQ("per_fmol", flatModel->units(1)->name());
```

Comment:
That is an interesting question.

Reply:
FWIW, I don't think libCellML should create something because the model author didn't create it in the first place. libCellML should just report issues, not "fix" them.

@nickerso, any views on this?

Reply:
I thought all units had to be defined for flattening to proceed? since the `unusual` units is undefined it can't be flattened. It definitely shouldn't be created!

I'm also wondering why there are 3 units in the flattened model, I'd only be expecting 2 - I think.

Reply:
I think there should only be two units also.
The current validity for flattening model is set around two criteria:

1. Does the model have any import issues
2. Is the model defined

The current logic says both of these criteria are true for this model.

Reply:
The model doesn't have any import issues, the undefined units `unusual` is not imported.
The model is well defined, the is defined test doesn't travel through imports testing for their definedness.

Reply:
I lie, `isDefined` does go through imports.  I'll dig a bit deeper.

Reply:
[48b777b](https://github.com/cellml/libcellml/pull/1167/commits/48b777be471781bea81a3d08e03465bd4a677fdd)

Reply:
`isDefined` wasn't checking that a units used by a variable on `cn` element was actually defined in the model, now it does and correctly determines that the model cannot be flattened due to ill-defined units.

Reply:
> `isDefined` wasn't checking that a units used by a variable on `cn` element was actually defined in the model, now it does and correctly determines that the model cannot be flattened due to ill-defined units.

Thanks, that's the behaviour I would indeed expect.